### PR TITLE
[FLINK-12610] [table-planner-blink] Introduce planner rules about aggregate

### PR DESCRIPTION
**Pull request by godfreyhe**
## What is the purpose of the change

*Introduce planner rules about aggregate*

## Brief change log

  - *AggregateReduceGroupingRule, that reduces unless grouping columns*
  - *PruneAggregateCallRule, that removes unreferenced AggregateCall from Aggregate*
  - *FlinkAggregateRemoveRule, that is copied from Calcite's AggregateRemoveRule, and supports SUM, MIN, MAX, AUXILIARY_GROUP functions in non-empty group aggregate*
  - *FlinkAggregateJoinTransposeRule, that is copied from Calcite's AggregateJoinTransposeRule, and supports Left/Right outer join and aggregate with AUXILIARY_GROUP*

## Verifying this change

This change added tests and can be verified as follows:

  - *Added rule tests that validate the logical plan after the added rules are applied*
  - *Added integration tests that validate the execution result the added rules are applied*

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / **not documented**)

### CODE REVIEW

File: flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/plan/rules/logical/AggregateCalcMergeRule.java [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * Planner rule that recognizes a {@link org.apache.calcite.rel.core.Aggregate}
+ * on top of a {@link org.apache.calcite.rel.core.Calc} and if possible
+ * aggregate through the calc or removes the calc.
+ *
+ * <p>This is only possible when no condition in calc and the grouping expressions and arguments to
```
Comment (KurtYoung) [inline flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/plan/rules/logical/AggregateCalcMergeRule.java]: If you can't merge condition, why not change the rule match pattern to `aggregate on project`?

File: flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/plan/rules/logical/FlinkAggregateJoinTransposeRule.java [hunk 2]
```diff
@@ -0,0 +1,593 @@
...
+import java.util.TreeMap;
+
+import scala.Tuple2;
+import scala.collection.JavaConverters;
+import scala.collection.Seq;
+
+/**
+ * This rule is copied from Calcite's {@link org.apache.calcite.rel.rules.AggregateJoinTransposeRule}.
+ * Modification:
+ * - Do not match TemporalTableScan since it means that it is a dimension table scan currently.
```
Comment (KurtYoung) [inline flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/plan/rules/logical/FlinkAggregateJoinTransposeRule.java]: Put this rule into CBO, and we can avoid to deal with such situation.

File: flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/plan/rules/logical/AggregateCalcMergeRule.java (hunk 1 above)
Comment (godfreyhe) [reply flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/plan/rules/logical/AggregateCalcMergeRule.java]: `AggregateProjectMergeRule` already exists and is in our rule set. The original intention of introducing this rule is to solve the hack in `RelDecorrelator` for tpch query20, however current cost model can not find best plan, so this rule has not been add to our rule set.  we can move this rule from this pr, and introduce it when needed

File: flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/plan/rules/logical/FlinkAggregateJoinTransposeRule.java (hunk 2 above)
Comment (godfreyhe) [reply flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/plan/rules/logical/FlinkAggregateJoinTransposeRule.java]: this rule is already in CBO, and CBO does not handle the unspported case (lookup table source doesn't support aggregate).

**Pull request merged**